compiler/util/ir_data.py

```diff
@@ -14,411 +14,143 @@
+  led to some legacy naming conventions and references. In particular this
+  class is named `Message` in the sense of a protocol buffer message,
+  indicating that it is intended to just be data that is used by other higher
+  level services.
 
-
-class Optional(object):
-  """Property implementation for "optional"-like fields."""
-
-  def __init__(self, type_, oneof=None, decode_names=None):
-    """Creates a Proto "optional"-like data member.
-
-    Args:
-      type_: The type of the field; e.g., _Int or _Text.
-      oneof: If set, the name of the proto-like "oneof" that this field is a
-          member of.  Within a structure, at most one field of a particular
-          "oneof" may be set at a time; setting a member of the "oneof" will
-          clear any other member that might be set.
-      decode_names: An optional callable that takes a str and returns a
-          value of type type_; allows strs to be used to set "enums" using
-          their symbolic names.
-    """
-    self._type = type_
-    self._oneof = oneof
-    self._decode_names = decode_names
-
-  def __get__(self, obj, type_=None):
-    result = obj.raw_fields.get(self.name, None)
-    if result is not None:
-      return result
-    if self.type in _BASIC_TYPES:
-      return self._type()
-    result = self._type()
-
-    def on_write():
-      self._set_value(obj, result)
-
-    result.set_on_write(on_write)
-    return result
-
-  def __set__(self, obj, value):
-    if issubclass(self._type, _BASIC_TYPES):
-      self.set(obj, value)
-    else:
-      raise AttributeError("Cannot set {} (type {}) for type {}".format(
-          value, value.__class__, self._type))
-
-  def _set_value(self, obj, value):
-    if self._oneof is not None:
-      current = obj.oneofs.get(self._oneof)
-      if current in obj.raw_fields:
-        del obj.raw_fields[current]
-      obj.oneofs[self._oneof] = self.name
-    obj.raw_fields[self.name] = value
-    obj.on_write()
-
-  def set(self, obj, value):
-    """Sets the given value to the property."""
-
-    if value is None:
-      return
-    if isinstance(value, dict):
-      self._set_value(obj, self._type(**value))
-    elif isinstance(value, _Text) and self._decode_names:
-      self._set_value(obj, self._type(self._decode_names(value)))
-    elif isinstance(value, _Text) and issubclass(self._type, enum.Enum):
-      self._set_value(obj, getattr(self._type, value))
-    elif (not isinstance(value, self._type) and
-          not (self._type == _Int and isinstance(value, _int_types)) and
-          not (self._type == _Text and isinstance(value, _text_types)) and
-          not (issubclass(self._type, enum.Enum) and isinstance(value, _int_types))):
-      raise AttributeError("Cannot set {} (type {}) for type {}".format(
-          value, value.__class__, self._type))
-    elif issubclass(self._type, Message):
-      self._set_value(obj, self._type(**value.raw_fields))
-    else:
-      self._set_value(obj, self._type(value))
```

Comment:
One minor concern I have with the whole `dataclass` approach is that it does not seem to enforce types in any way -- you can assign whatever value you want to any field.  We're not running any of the type checkers on the source (largely because there are very few type annotations), so that seems like it would be an easy bug to miss.

Reply:
So on the plus side our dataclasses now have type annotations and we could benefit from running type checkers. Additionally there's indication of misuse in IDEs that support type checking such as VS Code and I'd expect our tests to find issues (although perhaps not with as nice of a `AttributeError`).

Perhaps we could file a follow-up to enable type checking in our bazel build?

Reply:
To be clear: this would not block approval from me, and any large mitigation is outside the scope of this change (but should probably be captured in a GitHub issue).  That said, I think it would be useful to think about mitigations now.  My bug paranoia is that a wrong type *silently* makes it through to generated code and through C++ compilation: for example, I can imagine `""` working in some places that an `int` would work, like in `+ offset` instead of `16 + offset`.

One option is to add `__setattr__` and have it do type checking (maybe only while running tests).  Looking through `dataclasses.py`, it looks like a) `__setattr__` is not changed for non-frozen classes, and b) you could crib some of the logic in there to find the type annotations.

Running a static type checker wouldn't hurt, but I'm pretty sure none of them will be able to infer argument types for the functions that are passed into `traverse_ir`.  Someone would have to add annotations to those functions before a type checker could catch much of anything.

Reply:
Yep, I agree. Python itself as a `-O` option that disables assertions, so we could potentially just add in assertions / debug level code guarded by `__debug__` and avoid having a perf hit by running `embossc` via `python3 -O ./embossc`. This would work in some cases but might be difficult to specify via `bazel` integration. I'm not sure the best way to do this *just* for tests, we could possibly have a global config set via an env var but that has some overhead as well.

Reply:
I prototyped adding checks back guarded by `__debug__`, for a full run of `pw_bluetooth` I get:
```bash
ninja -C out/cmake_host pw_bluetooth.emboss_test.bin && \
multitime -n5 -r "rm -rf out/cmake_host/pw_bluetooth" \
    ninja -C out/cmake_host pw_bluetooth.emboss_test.bin

            Mean        Std.Dev.    Min         Median      Max
real        25.124      0.472       24.460      25.067      25.784
user        88.741      0.623       87.703      89.099      89.427
sys         2.286       0.132       2.135       2.294       2.506
```

And then if setup my builder to use `python3 -OO`:
```
ninja -C out/cmake_host pw_bluetooth.emboss_test.bin && \
multitime -n5 -r "rm -rf out/cmake_host/pw_bluetooth" \
    ninja -C out/cmake_host pw_bluetooth.emboss_test.bin

            Mean        Std.Dev.    Min         Median      Max
real        21.724      0.402       21.137      21.667      22.272
user        80.090      0.328       79.704      80.136      80.488
sys         2.203       0.098       2.110       2.169       2.367
```

So it's something like a 3.4s / 15% overhead. I think that's probably fine for now to give us some peace of mind and we can let users opt-in to running an optimized path.

We can file a follow-up to perhaps make this test-only instead, there are some options such as setting an env var, but that would require updating all of our tests.

Reply:
@reventlov added checks, PTAL